### ISSUE TITLE
Signal registration may lead to issues with multithreading

### ISSUE THREAD

**Issue by krassowski** [rpy2/rpy2#769]
**Describe the issue or bug**

The signal registration introduced in #736 can cause trouble when running in a thread/on a server. This will cause issues with Django, though some other servers will catch this exception (so it would work with Apache + WSGI but show a verbose warning in the logs) as discussed on [SO](https://stackoverflow.com/questions/3657732/how-to-import-the-python-async-module-from-a-worker-thread). This is a follow-up after #729.

It happens on `import rpy2.robjects`.

**To Reproduce**

```python
from rpy2 import rinterface

def rpy2_init_simple():
    rinterface.initr_simple()

def test_threading_simple():
    from threading import Thread

    thread = Thread(target=rpy2_init_simple)
    thread.start()

test_threading_simple()
```

**Expected behavior**

Should not attempt to register the signal when running in a thread. It would seem to be as simple as:

```python
if threading.current_thread() is threading.main_thread():
    signal.signal(signal.SIGINT, _sigint_handler)
```

but there is [this bug in Python 3.8+](https://bugs.python.org/issue38904) that might bite in certain circumstances, so it might be just ok to catch the `ValueError` and `pass` if and only if its value is `signal only works in main thread`

**Error**

WSGI:

```python
Callback registration for signal 2 ignored.
  File "/site-packages/rpy2/robjects/__init__.py", line 442, in <module>
    r = R()
  File "/site-packages/rpy2/robjects/__init__.py", line 402, in __new__
    rinterface.initr_simple()
  File "/site-packages/rpy2/rinterface.py", line 859, in initr_simple
    _post_initr_setup()
  File "/site-packages/rpy2/rinterface.py", line 940, in _post_initr_setup
    signal.signal(signal.SIGINT, _sigint_handler)
```

Less smart multi-threaded servers:

```python
  File "test.py", line 5, in rpy2_init_simple
    rinterface.initr_simple()
  File "/site-packages/rpy2/rinterface.py", line 859, in initr_simple
    _post_initr_setup()
  File "/site-packages/rpy2/rinterface.py", line 940, in _post_initr_setup
    signal.signal(signal.SIGINT, _sigint_handler)
  File "/lib/python3.7/signal.py", line 47, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
ValueError: signal only works in main thread
```

**Additional context**

- rpy2 3.4.2
- tested on Python 3.7.6 and Python 3.8.3

**Comment by lgautier** [comment]
What is happening to Ctrl-C when R is initialized in a thread that is not the main one then? Will there be a way to interrupt it?

**Comment by krassowski** [comment]
> What is happening to Ctrl-C when R is initialized in a thread that is not the main one then?

Nothing:

```
Skipping 2 <function _sigint_handler at 0x7ff28fae1e60>
^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C
```

after some 20 seconds it crash-exits:

```
R[write to console]: 

*** longjmp causes uninitialized stack frame ***: terminated
Aborted (core dumped)
```


```python
from rpy2 import rinterface
import signal
from time import sleep

def mock(signal_code, handler):
    print(f"Skipping {signal_code} {handler}")

signal.signal = mock

def rpy2_init_simple():
    rinterface.initr_simple()

def test_threading_simple():
    from threading import Thread

    thread = Thread(target=rpy2_init_simple)
    thread.start()

test_threading_simple()

for i in range(60):
    sleep(1)
```

So it does not exit gracefully, but at least it works. In perfect world the thread would be able to tell the main thread to hook the handler for the signal - I do not know if this is possible this way (it might be); the other way: main thread registers signal and nudges the thread to call _sigint_handler would certainly work, but might be useless in flask/django servers context as we do not control the main thread.

**Comment by lgautier** [comment]
I had a quick look at what is reported about using SIGINT in (Python) thread, and it seems that this is not too nice. We may just say that we are not making this substantially more problematic than it is already in Python and use the fix you suggest.

**Comment by lgautier** [comment]
Can someone have a look at PR #780 and tell whether it is solving the issue?

**Comment by krassowski** [comment]
I will look into it this week (probably closer to the weekend).

**Comment by lgautier** [comment]
> I will look into it this week (probably closer to the weekend).

Thanks @krassowski. Much appreciated. I am trying to see if I can fix one or two issues more to have a bugfix release soon.

**Comment by krassowski** [comment]
Looking into it now.

**Comment by lgautier** [comment]
The PR is merged. It will be included in rpy2-3.4.3 (that should be released today or tomorrow).
